crates/sui-indexer/src/handlers/checkpoint_handler_v2.rs

```diff
@@ -0,0 +1,635 @@
+                IndexedCheckpoint::from_sui_checkpoint(
+                    checkpoint_summary,
+                    checkpoint_contents,
+                    successful_tx_num as usize,
+                ),
+                db_transactions,
+                db_events,
+                db_indices,
+            )
+        };
+
+        let epoch = Self::index_epoch(state, &data).await?;
+
+        // Index Objects
+        let (object_changes, packages) =
+            Self::index_checkpoint(state, data, package_cache, metrics);
+
+        Ok(CheckpointDataToCommit {
+            checkpoint,
+            transactions: db_transactions,
+            events: db_events,
+            tx_indices: db_indices,
+            object_changes,
+            packages,
+            epoch,
+        })
+    }
+
+    fn index_checkpoint(
+        state: &S,
+        data: CheckpointData,
+        package_cache: Arc<Mutex<IndexingPackageCache>>,
+        metrics: &IndexerMetrics,
+    ) -> (TransactionObjectChangesToCommit, Vec<IndexedPackage>) {
+        let checkpoint_seq = data.checkpoint_summary.sequence_number;
+        info!(checkpoint_seq, "Indexing checkpoint");
+        let packages = Self::index_packages(&data, metrics);
+
+        let object_changes = Self::index_objects(state, data, &packages, package_cache, metrics);
+
+        (object_changes, packages)
+    }
+
+    fn index_objects(
+        state: &S,
+        data: CheckpointData,
+        packages: &[IndexedPackage],
+        package_cache: Arc<Mutex<IndexingPackageCache>>,
+        metrics: &IndexerMetrics,
+    ) -> TransactionObjectChangesToCommit {
+        let _timer = metrics.indexing_objects_latency.start_timer();
+        let checkpoint_seq = data.checkpoint_summary.sequence_number;
+        let module_resolver = InterimModuleResolver::new(
+            state.module_cache(),
+            package_cache,
+            packages,
+            checkpoint_seq,
+            metrics.clone(),
+        );
+        let deleted_objects = data
+            .transactions
+            .iter()
+            .flat_map(|tx| get_deleted_objects(&tx.effects))
+            .collect::<Vec<_>>();
+
+        let deleted_object_ids = deleted_objects
+            .iter()
+            .map(|o| (o.0, o.1))
+            .collect::<HashSet<_>>();
+
+        let (objects, intermediate_versions) = get_latest_objects(data.output_objects());
+
+        let changed_objects = data
+            .transactions
+            .iter()
+            .flat_map(|tx| {
+                let CheckpointTransaction {
+                    transaction: tx,
+                    effects: fx,
+                    ..
+                } = tx;
+                fx.all_changed_objects()
+                    .into_iter()
+                    .filter_map(|(oref, _owner, _kind)| {
+                        // We don't care about objects that are deleted or updated more than once
+                        if intermediate_versions.contains(&(oref.0, oref.1))
```

Comment:
iirc `output_objects` are from `all_changed_objects()`, in that sense filtering `intermediate_versions` seems not necessary?

Reply:
Imagine a checkpoint has 2 transactions. Transaction 1 updates object A to V1, Transaction 2 updates Object A to V2 (from V1). When we make prepare objects to commit, V1 will appear in `intermediate_versions` but there's no need to commit it.

Reply:
right I got the idea of skipping intermediate versions, I meant why not derive `changed_objects` from `objects` instead of `fx.all_changed_objects()`, as `objects` have filtered out intermediate versions already?

Reply:
I see what you are saying, this is to prevent the returned objects from fullnode is not complete for some reason, and we don't want to miss objects in that case. The `objects.get()` in line 474 does the check

Reply:
resolved offline